### ISSUE TITLE
Is this process continuous?

### ISSUE THREAD

**Issue by JinFish** [labmlai/annotated_deep_learning_paper_implementations#138]
# line 98
route_prob = self.softmax(self.switch(x))
# line 102
route_prob_max, routes = torch.max(route_prob, dim=-1)

As far as I know, the `torch.max` operation is discontinuous, so that the parameters of `self.switch` cannot be trained, which means that the experts will be chosen randomly each time, I don't know if there is something wrong with my understanding.

**Comment by vpj** [comment]
The gradients flow through `torch.max`.

e.g.
```python
t = torch.tensor([0., 1., 2., 3.], requires_grad=True)
m, _ = torch.max(t, dim=-1)
m.backward()
print(t.grad)
```

We [multiply the final output by `route_prob_max`](https://nn.labml.ai/transformers/switch/index.html#section-25). Therefore the gradients flow through it.

**Comment by JinFish** [comment]
Thank you very much for solving my problem and for getting back to me so quickly!
